beacon_node/beacon_chain/tests/store_tests.rs

```diff
@@ -1058,235 +1070,302 @@ fn prunes_fork_running_past_finalized_checkpoint() {
-
-    let skipped_slots = vec![canonical_slot, canonical_slot + 1];
-    for &slot in &skipped_slots {
-        assert_eq!(stray_state.get_block_root(slot), block_root);
-        let state_hash = stray_state.get_state_root(slot).unwrap();
+    assert!(rig.is_skipped_slot(&stray_state, skipped_slot));
+    {
+        let state_hash = (*stray_state.get_state_root(skipped_slot).unwrap()).into();
         assert!(
-            harness
-                .chain
-                .get_state(&state_hash, Some(slot))
-                .unwrap()
-                .is_some(),
-            "skipped slots state should be still present"
+            rig.hot_state_exists(state_hash),
+            "skipped slot state {} should be still present",
+            state_hash
         );
     }
 
     // Trigger finalization
-    let (canonical_blocks_post_finalization, _, _, _, _) = harness.add_canonical_chain_blocks(
+    let canonical_slots: Vec<Slot> = ((skipped_slot + 1).into()..rig.checkpoint_of_epoch(7))
+        .map(Into::into)
+        .collect();
+    let (canonical_blocks_post_finalization, _, _, _) = rig.add_attested_blocks_at_slots(
         canonical_state,
-        canonical_slot,
-        slots_per_epoch * 6,
-        &honest_validators,
+        &canonical_slots,
+        &rig.get_honest_validators(),
     );
+
+    // Postconditions
+    let canonical_blocks: HashMap<Slot, SignedBeaconBlockHash> = canonical_blocks_zeroth_epoch
+        .into_iter()
+        .chain(canonical_blocks_post_finalization.into_iter())
+        .collect();
     assert_eq!(
-        harness
-            .chain
-            .head()
-            .unwrap()
-            .beacon_state
-            .finalized_checkpoint
-            .epoch,
-        2,
-        "chain should have finalized"
+        rig.get_finalized_checkpoints(),
+        hashset! {
+            canonical_blocks[&rig.checkpoint_of_epoch(1).into()],
+            canonical_blocks[&rig.checkpoint_of_epoch(2).into()],
+        },
+    );
+
+    for (&slot, &state_hash) in &stray_states {
+        assert!(
+            !rig.hot_state_exists(state_hash),
+            "stray state {} at slot {} should have been pruned",
+            state_hash,
+            slot
+        );
+        assert!(
+            !rig.cold_state_exists(state_hash),
+            "stray state {} at slot {} should have been pruned",
+            state_hash,
+            slot
+        );
+    }
+
+    assert!(rig.is_skipped_slot(&stray_state, skipped_slot));
+    {
+        let state_hash: BeaconStateHash =
+            (*stray_state.get_state_root(skipped_slot).unwrap()).into();
+        assert!(
+            !rig.hot_state_exists(state_hash),
+            "skipped slot {} state {} should have been pruned",
+            skipped_slot,
+            state_hash
+        );
+    }
+}
+
+// This is to check if state outside of normal block processing are pruned correctly.
+#[test]
+#[should_panic]
```

Comment:
I think to deal with this more cleanly we should make the `BlockingMigrator` handle just the `FreezeSlotUnaligned` error (to stop it propagating to fork choice). This would make the `BlockingMigrator` more similar to the threaded one, which has a match especially for this case:

https://github.com/sigp/lighthouse/blob/master/beacon_node/beacon_chain/src/migrate.rs#L328-L344

If we want to assert that the split slot is distinct from the finalized checkpoint in tests, then we can check them directly with `db.get_split_slot()` and `head_state.finalized_checkpoint.epoch.start_slot(..)`

Reply:
Sounds good 👍